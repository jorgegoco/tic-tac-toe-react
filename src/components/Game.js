import React, { useState } from 'react';
import calculateWinner from '../helpers';
import Board from './Board';
import './game.css';

const Game = () => {
  const [history, setHistory] = useState([Array(9).fill(null)]);
  const [stepNumber, setStepNumber] = useState(0);
  const [xIsNext, setXIsNext] = useState(true);
  const winner = calculateWinner(history[stepNumber]);
  const winningSquares = winner ? calculateWinner(history[stepNumber]).winningSquares : [];
  const handleClick = (i) => {
    const timeInHistory = history.slice(0, stepNumber + 1);
    const current = timeInHistory[stepNumber];
    const squares = [...current];
    // If user clicks an occupied square or if game is won, return
    if (winner || squares[i]) return;
    // Put an X or an O in the clicked square
    squares[i] = xIsNext ? 'X' : 'O';
    setHistory([...timeInHistory, squares]);
    setXIsNext(!xIsNext);
    setStepNumber(timeInHistory.length);
  };

  const jumpTo = (step) => {
    if (step === 0) {
      // If step is 0, reset history, stepNumber, and xIsNext
      setHistory([Array(9).fill(null)]);
      setStepNumber(0);
      setXIsNext(true);
    } else {
      // Otherwise, update stepNumber and xIsNext based on the selected step
      setStepNumber(step);
      setXIsNext(step % 2 === 0);
    }
  };

  const renderMoves = () => (
    <ul className="game-moves">
      {history.map((_step, move) => {
        const destination = move ? `Go to move #${move}` : 'Go to start';
        return (
          // eslint-disable-next-line react/no-array-index-key
          <li key={move}>
            <button type="button" onClick={() => jumpTo(move)}>{destination}</button>
          </li>
        );
      })}
    </ul>
  );

  return (
    <div className="game">
      <h1 className="game-header">Tic Tac Toe</h1>
      <Board
        squares={history[stepNumber]}
        winningSquares={winningSquares}
        onClick={(i) => handleClick(i)}
      />
      <div>
        <p className="game-status">
          {winner
            ? `Winner: ${winner.winner}`
            : `Next Player: ${xIsNext ? 'X' : 'O'}`}
        </p>
        {renderMoves()}
      </div>
    </div>
  );
};

export default Game;
