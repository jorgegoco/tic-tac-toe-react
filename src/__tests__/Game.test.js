import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Game from '../components/Game';

test('renders Tic Tac Toe header', () => {
  const { getByText } = render(<Game />);
  const headerElement = getByText(/Tic Tac Toe/i);
  expect(headerElement).toBeInTheDocument();
});

test('renders game status with next player', () => {
  const { getByText } = render(<Game />);
  const statusElement = getByText(/Next Player: X/i);
  expect(statusElement).toBeInTheDocument();
});

test('clicking on a square updates the board', () => {
  const { getAllByRole } = render(<Game />);
  const squares = getAllByRole('button');
  fireEvent.click(squares[0]);
  expect(squares[0]).toHaveTextContent('X');
});

test('clicking on a square twice does not update the board', () => {
  const { getAllByRole } = render(<Game />);
  const squares = getAllByRole('button');
  fireEvent.click(squares[0]);
  fireEvent.click(squares[0]);
  expect(squares[0]).toHaveTextContent('X');
});

test('clicking on a winning square highlights the winning squares', () => {
  const { getAllByRole, getByTestId } = render(<Game />);
  const squares = getAllByRole('button');
  fireEvent.click(squares[0]);
  fireEvent.click(squares[3]);
  fireEvent.click(squares[1]);
  fireEvent.click(squares[4]);
  fireEvent.click(squares[2]);
  const boardElement = getByTestId('board');
  expect(boardElement).toMatchSnapshot();
});

test('clicking on a square after the game has been won does not update the board', () => {
  const { getAllByRole, getByText } = render(<Game />);
  const squares = getAllByRole('button');
  fireEvent.click(squares[0]);
  fireEvent.click(squares[3]);
  fireEvent.click(squares[1]);
  fireEvent.click(squares[4]);
  fireEvent.click(squares[2]);
  fireEvent.click(squares[5]);
  const statusElement = getByText(/Winner: X/i);
  expect(statusElement).toBeInTheDocument();
  expect(squares[5]).toHaveTextContent('');
});

test('clicking on the reset button resets the game board and status', () => {
  const { getAllByRole, getByText } = render(<Game />);
  const squares = getAllByRole('button');
  fireEvent.click(squares[0]);
  fireEvent.click(squares[3]);
  fireEvent.click(squares[1]);
  fireEvent.click(squares[4]);
  fireEvent.click(squares[2]);
  const resetButton = getByText(/Start/i);
  fireEvent.click(resetButton);
  expect(squares[0]).toHaveTextContent('');
  expect(squares[1]).toHaveTextContent('');
  expect(squares[2]).toHaveTextContent('');
  expect(squares[3]).toHaveTextContent('');
  expect(squares[4]).toHaveTextContent('');
  expect(squares[5]).toHaveTextContent('');
  expect(squares[6]).toHaveTextContent('');
  expect(squares[7]).toHaveTextContent('');
  expect(squares[8]).toHaveTextContent('');
  const statusElement = getByText(/Next Player: X/i);
  expect(statusElement).toBeInTheDocument();
});

test('clicking on a "Go to move #" button updates the board to the correct move', () => {
  const { getAllByRole, getByText } = render(<Game />);
  const squares = getAllByRole('button');
  fireEvent.click(squares[0]);
  fireEvent.click(squares[3]);
  fireEvent.click(squares[1]);
  fireEvent.click(squares[4]);
  fireEvent.click(squares[2]);
  const moveButton = getByText(/Go to move #3/i);
  fireEvent.click(moveButton);
  expect(squares[0]).toHaveTextContent('X');
  expect(squares[1]).toHaveTextContent('X');
  expect(squares[2]).toHaveTextContent('');
  expect(squares[3]).toHaveTextContent('O');
  expect(squares[4]).toHaveTextContent('');
  expect(squares[5]).toHaveTextContent('');
  expect(squares[6]).toHaveTextContent('');
  expect(squares[7]).toHaveTextContent('');
  expect(squares[8]).toHaveTextContent('');
});

test('clicking on the "Start" button resets the game board and status', () => {
  const { getAllByRole, getByText } = render(<Game />);
  const squares = getAllByRole('button');
  fireEvent.click(squares[0]);
  fireEvent.click(squares[3]);
  fireEvent.click(squares[1]);
  fireEvent.click(squares[4]);
  fireEvent.click(squares[2]);
  const startButton = getByText(/Start/i);
  fireEvent.click(startButton);
  expect(squares[0]).toHaveTextContent('');
  expect(squares[1]).toHaveTextContent('');
  expect(squares[2]).toHaveTextContent('');
  expect(squares[3]).toHaveTextContent('');
  expect(squares[4]).toHaveTextContent('');
  expect(squares[5]).toHaveTextContent('');
  expect(squares[6]).toHaveTextContent('');
  expect(squares[7]).toHaveTextContent('');
  expect(squares[8]).toHaveTextContent('');
  const statusElement = getByText(/Next Player: X/i);
  expect(statusElement).toBeInTheDocument();
});
