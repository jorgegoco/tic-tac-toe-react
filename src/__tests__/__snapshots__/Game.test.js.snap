// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clicking on a winning square highlights the winning squares 1`] = `
<div
  class="board"
  data-testid="board"
>
  <button
    class="square winning"
    type="button"
  >
    X
  </button>
  <button
    class="square winning"
    type="button"
  >
    X
  </button>
  <button
    class="square winning"
    type="button"
  >
    X
  </button>
  <button
    class="square "
    type="button"
  >
    O
  </button>
  <button
    class="square "
    type="button"
  >
    O
  </button>
  <button
    class="square "
    type="button"
  />
  <button
    class="square "
    type="button"
  />
  <button
    class="square "
    type="button"
  />
  <button
    class="square "
    type="button"
  />
</div>
`;
